datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,282 @@
+import os
+from typing import Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkMetadata,
+    DocumentMetadataFilter,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding,
+)
+import json
+import psycopg
+from psycopg import sql
+from psycopg_pool import ConnectionPool
+
+from services.date import to_unix_timestamp
+
+# Read environment variables for Postgres
+
+POSTGRES_HOST = os.environ.get("POSTGRES_HOST", "localhost")
+POSTGRES_PORT = int(os.environ.get("POSTGRES_PORT", 5433))
+POSTGRES_USERNAME = os.environ.get("POSTGRES_USERNAME", "postgres")
+POSTGRES_PASSWORD = os.environ.get("POSTGRES_PASSWORD", "")
+POSTGRES_DATABASE = os.environ.get("POSTGRES_DATABASE", "pgml_development")
+POSTGRES_TABLENAME = os.environ.get("POSTGRES_TABLENAME", "chatgpt_datastore")
+POSTGRES_UPSERT_COMMAND = os.environ.get("POSTGRES_UPSERT_COMMAND", "INSERT")
+
+assert POSTGRES_UPSERT_COMMAND in ["COPY", "INSERT"]
+
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+
+
+class PostgresDataStore(DataStore):
+    def __init__(self) -> None:
+        conninfo = (
+            "postgresql://"
+            + POSTGRES_USERNAME
+            + ":"
+            + POSTGRES_PASSWORD
+            + "@"
+            + POSTGRES_HOST
+            + ":"
+            + str(POSTGRES_PORT)
+            + "/"
+            + POSTGRES_DATABASE
+        )
+        self.pool = ConnectionPool(conninfo)
+        conn = self.pool.getconn()
+        conn.autocommit = True
+        # Insert the vector and text into the database
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
```

Comment:
is chunkid ordinal?.. unrelated to this PR, just curious

Reply:
It is ordinal. Text is first converted to tokens using tiktoken `tiktoken.get_encoding("cl100k_base")`, tokens are chunked based on chunk size (default 100), tokens are converted back to text. Each chunk gets an id with doc_id as a prefix. 

|doc_id | chunk_id|
|---------|-----------|
|5733be284776f41900661182|	5733be284776f41900661182_0|
|5733bf84d058e614000b61be|	5733bf84d058e614000b61be_0|
|5733bf84d058e614000b61be|	5733bf84d058e614000b61be_1|
|5733bed24776f41900661188|	5733bed24776f41900661188_0|
|5733a7bd4776f41900660f6b|	5733a7bd4776f41900660f6b_0|
|5733a7bd4776f41900660f6b|	5733a7bd4776f41900660f6b_1|